test/common/stats/allocator_impl_test.cc

```diff
@@ -75,6 +77,86 @@ TEST_F(AllocatorImplTest, GaugesWithSameName) {
+// iterate 10k times.
+TEST_F(AllocatorImplTest, RefCountDecAllocRaceOrganic) {
+  StatName counter_name = makeStat("counter.name");
+  StatName gauge_name = makeStat("gauge.name");
+  Thread::ThreadFactory& thread_factory = Thread::threadFactoryForTest();
+
+  const uint32_t num_threads = 12;
+  const uint32_t iters = 10000;
+  std::vector<Thread::ThreadPtr> threads;
+  absl::Notification go;
+  for (uint32_t i = 0; i < num_threads; ++i) {
+    threads.push_back(thread_factory.createThread([&]() {
+      go.WaitForNotification();
+      for (uint32_t i = 0; i < iters; ++i) {
+        alloc_.makeCounter(counter_name, "", std::vector<Tag>());
+        alloc_.makeGauge(gauge_name, "", std::vector<Tag>(), Gauge::ImportMode::NeverImport);
+      }
+    }));
+  }
+  go.Notify();
+  for (uint32_t i = 0; i < num_threads; ++i) {
+    threads[i]->join();
+  }
+}
+
+// Tests the same scenario as RefCountDecAllocRaceOrganic, but using just two
+// threads and the ThreadSynchronizer, in one iteration. Note that if the code
+// has the bug in it, this test fails fast as expected. However, if the bug is
+// fixed, the allocator's mutex will cause the second thread to block in
+// makeCounter() until the first thread finishes destructing the object. Thus
+// the test gives thread2 5 seconds to complete before releasing thread 1 to
+// complete its destruction of the counter.
+TEST_F(AllocatorImplTest, RefCountDecAllocRaceSynchronized) {
+  StatName counter_name = makeStat("counter.name");
+  Thread::ThreadFactory& thread_factory = Thread::threadFactoryForTest();
+  alloc_.sync().enable();
+  alloc_.sync().waitOn(AllocatorImpl::DecrementToZeroSyncPoint);
+  absl::Notification alloc_done;
+  Thread::ThreadPtr thread1 = thread_factory.createThread([&]() {
+    CounterSharedPtr counter1 = alloc_.makeCounter(counter_name, "", std::vector<Tag>());
+    alloc_done.Notify();
+    counter1->inc();
+    counter1->reset(); // Blocks in thread synchronizer waiting on DecrementToZeroSyncPoint
+  });
+
+  alloc_done.WaitForNotification();
+
+  absl::Mutex counter2_created_mutex;
+  bool counter2_created = false;
+
+  // counter1 has now been allocated in the thread, and is now in the middle of
+  // destructing it.
+  Thread::ThreadPtr thread2 = thread_factory.createThread([&]() {
+    CounterSharedPtr counter2 = alloc_.makeCounter(counter_name, "", std::vector<Tag>());
+    {
+      absl::MutexLock lock(&counter2_created_mutex);
+      counter2_created = true;
+    }
+    counter2->inc();
+
+    // We test for a value of 1 here to show that the first instance of the
+    // counter was destructed prior to the second instance being created, and
+    // thus starts again from zero.
+    EXPECT_EQ(1, counter2->value());
+  });
+
+  {
+    absl::MutexLock lock(&counter2_created_mutex);
+    counter2_created_mutex.AwaitWithTimeout(absl::Condition(&counter2_created), absl::Seconds(5));
+    EXPECT_FALSE(counter2_created);
```

Comment:
Just brainstorming a less hacky way to do this. Could we somehow actually assert/verify that the lock is already held? It seems bad that we need to wait 5 real seconds for this test to complete, though nothing immediately comes to mind on how to better test it. Having read through this again, I do feel that this synchronizer test adds value as its very clear what is being tested (as an added on to the brute force test above). WDYT?

Reply:
Yes I can do that. In fact I have that locally. Even then I will need some kind of sleep or other way to block the test's main thread until thread1 reaches the code inside `counter->reset()`, so I am not sure how this helps. If I don't sleep the test fails about 350 times out of 1000. With a 1 second sleep it works all the time.

I think the test as is still shows a lot by making the timeout 1 second rather than 5. But I still feel more confident including the brute force test as well. I do agree that it is testing the contention issue more directly though so Im fine leaving it in.

I propose I change the 5 seconds to 1 and we go forward from there?

Reply:
Couldn't you just have another sync point inside the code you need to reach?

Reply:
actually the sync-point I have works; I just needed to use barrierOn() rather than sleep(...); done.